### PR TITLE
Add aws client caching support

### PR DESCRIPTION
**Pull request by iomodo**
#### Summary
This PR will add AWS client caching, also support of the Mattermost cloud. User can call 
`/apps experiment --app "appname" --release "http://github.com/....zip"` 
to install app in the Mattermost cloud. 

User should pass `keyID` and `secret` to install app in the 3rd party environment.

### CODE REVIEW

File: server/command/install.go [hunk 1]
```diff
@@ -91,23 +87,14 @@ func (s *service) executeExperimentalInstall(params *params) (*model.CommandResp
...
-		return normalOut(params, nil, err)
+	if !s.apps.AWSProxy.IsAppInstalled(app) {
+		if err := s.apps.AWSProxy.InstallApp(app, releaseURL, id, secret); err != nil {
+			return normalOut(params, nil, err)
+		}
 	}
+
 	// An example of the function Invoke
-	// res, err := client.InvokeFunction("my_app_my_func", "blabla")
+	// res, err := s.apps.AWSProxy.InvokeFunction("my_app", "my_func", "blabla")
```
Comment (larkox) [inline server/command/install.go:97]: nit: Maybe change `blabla` for something like `my values`?

File: server/aws/credential_store.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	if len(value) == 0 {
+		ok, err := s.Mattermost.KV.Set(appsKey, &value)
+		if err != nil {
+			return errors.Wrapf(err, "can't set app list in KV store")
+		}
+		if !ok {
+			return errors.Errorf("can't set apps list in KV store")
+		}
+	}
+	return nil
```
Comment (larkox) [inline server/aws/credential_store.go:52]: Is this process needed? We are getting a value from the KVStore, which if it does not exist, will return a empty list of string, and we initialize it with a empty list of strings. Isn't it a bit redundant?

File: server/aws/credential_store.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	if !ok {
+		return errors.Errorf("can't set app %s credential from KV store", appID)
+	}
+
+	appsKey := getAppsKey()
+	var appsList []string
+	if err = s.Mattermost.KV.Get(appsKey, &appsList); err != nil {
+		return errors.Wrap(err, "can't get apps list from KV store")
+	}
+	appsList = append(appsList, appID)
```
Comment (larkox) [inline server/aws/credential_store.go:83]: Should we check if the app exist already in the list? Or is it handled somewhere else?

File: server/aws/credential_store.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		return errors.Wrapf(err, "can't set app %s credential from KV store", appID)
+	}
+	if !ok {
+		return errors.Errorf("can't set app %s credential from KV store", appID)
+	}
+
+	appsKey := getAppsKey()
+	var appsList []string
+	if err = s.Mattermost.KV.Get(appsKey, &appsList); err != nil {
+		return errors.Wrap(err, "can't get apps list from KV store")
```
Comment (larkox) [inline server/aws/credential_store.go:81]: Is there any big problem if the credentials are stored but the app is not added to the list? If there is, we should at least try here (and other early returns) to delete the credentials.

File: server/aws/credential_store.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	ok, err := s.Mattermost.KV.Set(appsKey, &appsList)
+	if err != nil {
+		return errors.Wrap(err, "can't set apps list in KV store")
+	}
+	if !ok {
+		return errors.Errorf("can't set apps list in KV store")
+	}
+	return nil
```
Comment (larkox) [inline server/aws/credential_store.go:120]: I wonder if we should try to be atomic with these updates on the list of apps. Set with options should allow atomic functionality.

File: server/aws/client_cache.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get credentials")
+	}
+	cc := &clientCache{
+		store: store,
+		cache: cache,
+	}
+	for _, appID := range list {
+		if _, err := cc.create(appID); err != nil {
+			return nil, errors.Wrapf(err, "can't create client for app %s", appID)
```
Comment (larkox) [inline server/aws/client_cache.go:33]: Should this be an return or a log + continue?

File: server/aws/client_cache.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	list, err := store.GetAppList()
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get credentials")
+	}
+	cc := &clientCache{
+		store: store,
+		cache: cache,
+	}
+	for _, appID := range list {
+		if _, err := cc.create(appID); err != nil {
```
Comment (larkox) [inline server/aws/client_cache.go:32]: Is it better to create all the clients on init, or it is better to create the clients on demand? (i.e. in the get function if the client does not exist)

File: server/aws/client_cache.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+// DefaultRegion describes default region in aws
+const DefaultRegion = "us-east-2"
+
+type clientCache struct {
+	store CredentialStoreService
+	cache map[string]*Client // appID - Client
+}
+
+func newClientCache(store CredentialStoreService) (*clientCache, error) {
```
Comment (larkox) [inline server/aws/client_cache.go:21]: How does the cache gets updated / invalidated? For example, when an app gets added/deleted. Not sure if the apps will change the credentials at any moment.

**Review by larkox — CHANGES_REQUESTED** [review]
Lots of questions, and a few nits here and there.

**Review by levb — CHANGES_REQUESTED** [review]
Can we start with a much simpler implementation of having a single set of AWS credentials stored in the (MMinstance/plugin-apps) config?

2/5 Using KV store we'll run into the usual performance/can't cache issues.

**Pull request closed without merging**